scm/build/Dockerfile

```diff
@@ -1,24 +1,139 @@
-FROM alpine:latest
+#
+# Stage: Build container. Used to build all binaries, once
+#
+FROM alpine:latest AS build-deps
 MAINTAINER Levente Kale <levente.kale@nokia.com>
 
-ENV GOPATH=/usr/local/go
-ENV PATH $GOPATH/bin:/usr/local/go/bin:$PATH
+ENV GOPATH=/go
+ENV GOROOT=/usr/local/go
+ENV PATH ${GOPATH}/bin:/usr/local/go/bin:$PATH
 ENV GO_VERSION 1.12.9
-
-COPY build.sh /build.sh
+ENV GOOS linux
+ENV CGO_ENABLED 0
+ENV GOARCH amd64
 
 RUN apk add --no-cache ca-certificates \
  && apk update --no-cache \
  && apk upgrade --no-cache \
  && apk add --no-cache make gcc musl-dev glide git bash curl tar \
  && mkdir -p ${GOPATH} \
+ && mkdir -p ${GOROOT} \
  && mkdir -p ${GOPATH}/bin \
  && mkdir -p ${GOPATH}/src \
- && curl -fsSL -k https://dl.google.com/go/go${GO_VERSION}.src.tar.gz | tar zx --strip-components=1 -C ${GOPATH} \
- && cd ${GOPATH}/src/ \
+ && curl -fsSL -k https://dl.google.com/go/go${GO_VERSION}.src.tar.gz | tar zx --strip-components=1 -C ${GOROOT} \
+ && cd ${GOROOT}/src/ \
  && ./make.bash \
  && rm -rf /var/cache/apk/* \
  && rm -rf /var/lib/apt/lists/* \
  && rm -rf /tmp/*
 
-ENTRYPOINT /build.sh
+RUN mkdir -p ${GOPATH}/src/github.com/nokia/danm
+COPY glide.yaml glide.lock ${GOPATH}/src/github.com/nokia/danm/
+COPY scm/build/build-deps.sh /
+
+WORKDIR ${GOPATH}/src/github.com/nokia/danm
+
+RUN /build-deps.sh
+
+
+#
+# Stage: Build actual binaries. This builder is used to build all DANM
+# related binaries.
+#
+# It is separated from the "build-deps" stage because mostly because
+# building the dependencies takes quite a lot of time and they (presumably)
+# don't change often -- so this could be factored out into a reusable
+# base image.
+#
+FROM build-deps AS builder
+
+COPY / ${GOPATH}/src/github.com/nokia/danm/
+
+ARG COMMIT_HASH
+ARG LATEST_TAG
+
+RUN scm/build/build.sh
+
+
+#
+# Stage: Netwatcher
+#
+FROM alpine:latest AS netwatcher
+MAINTAINER Levente Kale <levente.kale@nokia.com>
+
+COPY --from=builder /go/bin/netwatcher /usr/local/bin/netwatcher
+
+RUN apk add --no-cache --virtual .tools curl libcap iputils  \
+ && adduser -u 147 -D -H -s /sbin/nologin danm \
+ && chown root:danm /usr/local/bin/netwatcher \
```

Comment:
What about doing `chown` in `COPY` like `COPY --from=builder --chown root:danm ...`
Also you have to do `adduser` before `COPY` to work

Reply:
Quite honestly I pretty much just moved what was in the previous component Dockerfiles (eg. `integration/dockerfiles/netwatcher/Dockerfile` in this example) and used it 1:1. But I do agree that this might be an opportunity to review these tasks. Will modify as you suggested...

Reply:
Ok, cleaned up the Docker stage builds a bit... probably best to re-review the entire file as a whole :)

In short,

- moved all the chown/chmod into the build stage;
- created an intermitted stage which is just the Alpine image and adds the user (and then acts as a parent stage for netwatcher, svcwatcher, webhook)
- removed arping,
- removed a whole bunch of apk-add/apk-del that seemed to no longer serve any purpose